dig_test.go

```diff
@@ -0,0 +1,360 @@
+		c := New()
+		require.NoError(t, c.Provide(func() (*bytes.Buffer, error) {
+			return &bytes.Buffer{}, nil
+		}), "provide failed")
+		require.NoError(t, c.Invoke(func(b *bytes.Buffer) {
+			require.NotNil(t, b, "invoke got nil buffer")
+		}), "invoke failed")
+	})
+	t.Run("constructor fails", func(t *testing.T) {
+		c := New()
+		require.NoError(t, c.Provide(func() (*bytes.Buffer, error) {
+			return nil, errors.New("oh no")
+		}), "provide failed")
+
+		var called bool
+		err := c.Invoke(func(b *bytes.Buffer) { called = true })
+		assert.Contains(t, err.Error(), "oh no", "expected to bubble up constructor error")
+		assert.False(t, called, "shouldn't call invoked function when deps aren't available")
+	})
+}
+
+func TestCantProvideUntypedNil(t *testing.T) {
+	t.Fatalf("Calling Container.Provide(nil) crashes the main testing thread. Failing early.")
+	t.Parallel()
+	c := New()
+	assert.Error(t, c.Provide(nil))
+}
+
+func TestCantProvideErrors(t *testing.T) {
+	t.Parallel()
+	c := New()
+
+	assert.Error(t, c.Provide(func() error { return errors.New("foo") }))
+	// TODO: This is another case where we're losing type information, which
+	// (again) makes the provide-instance path behave differently from the
+	// provide-constructor path. This is fixable by not allowing users to
+	// provide types that *implement* error, but let's discuss a holistic
+	// solution.
+	assert.NoError(t, c.Provide(errors.New("foo")))
+}
+
+func TestProvideKnownTypesFails(t *testing.T) {
+	t.Parallel()
+	c := New()
+	assert.NoError(t, c.Provide(func() *bytes.Buffer { return nil }))
+	assert.Error(t, c.Provide(func() *bytes.Buffer { return nil }))
+}
+
+func TestProvideCycleFails(t *testing.T) {
+	t.Parallel()
+
+	// A <- B <- C
+	// |         ^
+	// |_________|
+	type A struct{}
+	type B struct{}
+	type C struct{}
+	newA := func(*C) *A { return &A{} }
+	newB := func(*A) *B { return &B{} }
+	newC := func(*B) *C { return &C{} }
+
+	c := New()
+	assert.NoError(t, c.Provide(newA))
+	assert.NoError(t, c.Provide(newB))
+	err := c.Provide(newC)
+	require.Error(t, err, "expected error when introducing cycle")
+	require.Contains(t, err.Error(), "cycle")
+}
+
+func TestIncompleteGraphIsOkay(t *testing.T) {
+	t.Parallel()
+
+	// A <- B <- C
+	// Even if we don't provide B, we should be able to resolve A.
+	type A struct{}
+	type B struct{}
+	type C struct{}
+	newA := func() *A { return &A{} }
+	newC := func(*B) *C { return &C{} }
+
+	c := New()
+	assert.NoError(t, c.Provide(newA), "provide failed")
+	assert.NoError(t, c.Provide(newC), "provide failed")
+	assert.NoError(t, c.Invoke(func(*A) {}), "invoke failed")
+}
+
+func TestProvideFuncsWithoutReturnsFails(t *testing.T) {
+	t.Parallel()
+
+	c := New()
+	assert.Error(t, c.Provide(func(*bytes.Buffer) {}))
+}
+
+func TestInvokesUseCachedObjects(t *testing.T) {
+	t.Parallel()
+
+	c := New()
+	calls := 0
+	buf := &bytes.Buffer{}
+	require.NoError(t, c.Provide(func() *bytes.Buffer { return buf }))
```

Comment:
as an extra safety, can we have the provided `func` increment a counter and ensure that the value is 0 initially, and then 1 the loop